fem/particleset.cpp

```diff
@@ -0,0 +1,902 @@
+                 GetEmptyNameArray(num_tags))
+{
+
+}
+
+ParticleSet::ParticleSet(MPI_Comm comm_, int rank_num_particles, int dim,
+                         const Array<int> &field_vdims, const Array<const
+                         char*> &field_names_,
+                         int num_tags, const Array<const char*> &tag_names_,
+                         Ordering::Type all_ordering)
+   : ParticleSet(comm_, rank_num_particles, dim, all_ordering, field_vdims,
+                 GetOrderingArray(all_ordering, field_vdims.Size()),
+                 field_names_, num_tags,
+                 tag_names_)
+{
+
+}
+
+ParticleSet::ParticleSet(MPI_Comm comm_, int rank_num_particles, int dim,
+                         Ordering::Type coords_ordering,
+                         const Array<int> &field_vdims,
+                         const Array<Ordering::Type> &field_orderings,
+                         const Array<const char*> &field_names_, int num_tags,
+                         const Array<const char*> &tag_names_)
+   : ParticleSet(GetSize(comm_), GetRank(comm_),
+                 rank_num_particles,
+                 dim,
+                 coords_ordering,
+                 field_vdims,
+                 field_orderings,
+                 field_names_,
+                 num_tags,
+                 tag_names_)
+{
+   comm = comm_;
+#ifdef MFEM_USE_GSLIB
+   gsl_comm = std::make_unique<gslib::comm>();
+   cr = std::make_unique<gslib::crystal>();
+   comm_init(gsl_comm.get(), comm);
+   crystal_init(cr.get(), gsl_comm.get());
+#endif // MFEM_USE_GSLIB
+}
+#endif // MFEM_USE_MPI
+
+unsigned int ParticleSet::GetGlobalNParticles() const
+{
+   unsigned int total = GetNParticles();
+#ifdef MFEM_USE_MPI
+   MPI_Allreduce(MPI_IN_PLACE, &total, 1, MPI_UNSIGNED, MPI_SUM, comm);
+#endif // MFEM_USE_MPI
+   return total;
+}
+
+int ParticleSet::AddField(int vdim, Ordering::Type field_ordering,
+                          const char* field_name)
+{
+   std::string field_name_str(field_name ? field_name : "");
+   if (!field_name)
+   {
+      field_name_str = GetDefaultFieldName(field_names.size());
+   }
+   fields.emplace_back(std::make_unique<ParticleVector>(vdim, field_ordering,
+                                                        GetNParticles()));
+   field_names.emplace_back(field_name_str);
+
+   return GetNFields() - 1;
+}
+
+int ParticleSet::AddTag(const char* tag_name)
+{
+   std::string tag_name_str(tag_name ? tag_name : "");
+   if (!tag_name)
+   {
+      tag_name_str = GetDefaultTagName(tag_names.size());
+   }
+   tags.emplace_back(std::make_unique<Array<int>>(GetNParticles()));
+   tag_names.emplace_back(tag_name_str);
+
+   return GetNTags() - 1;
+}
+
+void ParticleSet::AddParticle(const Particle &p)
+{
+   MFEM_ASSERT(IsValidParticle(p),
+               "Particle is incompatible with ParticleSet.");
+
+   // Add the particle
+   Array<int> idxs;
+   AddParticles(Array<int>({id_counter}), &idxs);
+   id_counter += id_stride;
```

Comment:
I would add a small private method for generation of unique ids, centralizing this striding logic.